app/controllers/omniauth_callbacks_controller.rb

```diff
@@ -0,0 +1,95 @@
+#
+# Handle signup or signin
+# from various oauth providers
+#
+# Heavily overlaps with Authentications controller
+#
+class OmniauthCallbacksController < Devise::OmniauthCallbacksController
+  def facebook
+    create
+  end
+  def failure
+    flash[:error] = "Authentication failed."
+    redirect_to request.env['omniauth.origin'] || "/"
+  end
+
+  private
+
+  #
+  # Inspects the omniauth information
+  # and determines if we have an existing member 
+  # (to add authentication to)
+  # or if this is a new signup
+  #
+  # As a side affect, this method sets the @member_created
+  # variable
+  #
+  def member
+    @member_created = false
+
+    auth = request.env['omniauth.auth']
+
+    authentication = Authentication.where(provider: auth.provider, uid: auth.uid).first
+    if authentication && authentication.member.id.present?
+      member = authentication.member
+    end
+
+    member = Member.where(email: auth.info.email).first_or_create do |m|
+      m.email = auth.info.email
+      m.password = Devise.friendly_token[0,20]
+      m.tos_agreement = true
+      # TODO This has a reasonable chance of collision
+      m.login_name = auth.info.nickname || auth.info.email.split("@").first.gsub(/[^A-Za-z]+/, '_').underscore
+
+      m.skip_confirmation!
```

Comment:
Just noting that we should think about this, if we're happy to skip email confirmation since facebook will have already done it, or if we still want to do it. I think I probably agree with the code here that we should skip it. It might be worth adding a comment to remind us of this reasoning in future.